engine/resources/hetzner_server.go

```diff
@@ -0,0 +1,804 @@
+// Mgmt
+// Copyright (C) 2013-2021+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+	"github.com/purpleidea/mgmt/util/errwrap"
+
+	"github.com/hetznercloud/hcloud-go/hcloud"
+)
+
+const (
+	// HetznerPollLimit sets a lower limit on polling interval in seconds.
+	// Note that the hetzner API supports bursts at up to 3600 requests per hour.
+	// A higher polling interval is recommended when expecting high rates
+	// of operational http requests to the hetzner api.
+	HetznerPollLimit = 1
+
+	// HetznerWaitIntervalLimit sets a lower limit on wait intervals in seconds.
+	// Nonzero values disable burst requests to manage rate limit errors.
+	// A wait interval greater than HetznerPollLimit is recommended.
+	HetznerWaitIntervalLimit = 0
+
+	// HetznerWaitTimeoutDefault sets a default timeout limit in seconds.
+	HetznerWaitTimeoutDefault = 5 * 60
+
+	// StateRunning indicates the server is powered on.
+	StateRunning = "running"
+	// StateOff indicates the server is turned off.
+	StateOff = "off"
+	// StateAbsent indicates the server does not exist at this time.
+	StateAbsent = "absent"
+
+	// ServerRescueDisabled indicates that rescue mode should be disabled.
+	ServerRescueDisabled = ""
+	// ServerRescueTypeLinux32 specifies the image to be used in rescue mode.
+	ServerRescueTypeLinux32 = "linux32"
+	// ServerRescueTypeLinux64 specifies the image to be used in rescue mode.
+	ServerRescueTypeLinux64 = "linux64"
+	// ServerRescueTypeFreeBSD64 specifies the image to be used in rescue mode.
+	ServerRescueTypeFreeBSD64 = "freebsd64"
+)
+
+func init() {
+	engine.RegisterResource("hetzner:server", func() engine.Res { return &HetznerServerRes{} })
```

Comment:
I would propose using a more specific resource name, maybe
"hetzner:cloud:vm" (personal preference) or "hetzner:vm"

Hetzner operates 2 lines of product lines: physical servers and virtual machines plus some surrounding services for each. Both product lines have their own API (hcloud [1] for VMs and robot [2] for physical machines) which AFAIK are custom-made.

My concern is that a too generic name for this resource might conflict with the implementation of future features, therefore my suggestion to change the resource name to something more specific.

@purpleidea 
I assume that naming of MCL resource types and their attributes should be static over mgmts lifetime to prevent breaking changes in the language. Is that a fair assumption?

[1] https://docs.hetzner.cloud/
[2] https://robot.your-server.de/doc/webservice/en.html

Reply:
>  "hetznercloudvm" (personal preference) or "hetzner:vm"
heh, the latter is fine.

>  I assume that naming of MCL resource types and their attributes should be static over mgmts lifetime to prevent breaking changes in the language. Is that a fair assumption?

This is fair! We'll try and keep everything static once merged, although I urge anyone to recommend any changes or renames of either resource names, param names, or any behaviours before 1.0 or perhaps 0.1 as that will be when we provide an API guarantee.

Reply:
If we really have to maintain consistency across the mgmt language then I think it's also important to consider the same naming problem for other cloud providers that might be added in the future. If we want to add eg. OVH support at some point, maybe it's cleaner to turn the names around and work with server:hcloud, server:robot, server:ovh...?

This way you define the function first, and then the name of (one of) the supported provider(s). Thoughts? I do realize that this implies that the API token for example should be called token:hcloud, and in that case the token specifier might not be super clear to the user...